sw/device/lib/dif/dif_pwm.h

```diff
@@ -0,0 +1,322 @@
+   * "beats" / "pulse cycle").
+   */
+  uint32_t beats_per_pulse_cycle;
+} dif_pwm_config_t;
+
+/**
+ * A PWM channel mode.
+ */
+typedef enum dif_pwm_mode {
+  /**
+   * The PWM duty cycle is set by the firmware and remains constant.
+   */
+  kDifPwmModeFirmware = 0,
+  /**
+   * The PWM duty cycle linearly sweeps between both primary and secondary
+   * firmware-configured values, based on a firmware-configured step size.
+   */
+  kDifPwmModeHeartbeat = 1,
+  /**
+   * The PWM duty cycle alternates between both primary and secondary
+   * firmware-configured values, based on two firmware-configured durations.
+   */
+  kDifPwmModeBlink = 1,
+} dif_pwm_polarity_t;
+
+/**
+ * A PWM channel polarity.
+ */
+typedef enum dif_pwm_polarity {
+  /**
+   * A PWM signal is active-high.
+   */
+  kDifPwmActiveHigh = 0,
+  /**
+   * A PWM signal is active-low.
+   */
+  kDifPwmActiveLow = 1,
+} dif_pwm_polarity_t;
+
+/**
+ * Runtime configuration for a specific PWM channel.
+ *
+ * This struct describes runtime configuration for one-time configuration of a
+ * specific PWM channel.
+ */
+typedef struct dif_pwm_channel_config {
+  /**
+   * Phase delay at the beginning of a "pulse cycle" to delay the active
+   * duty cycle "beats" for.
+   */
+  uint16_t phase_delay;
+  /**
+   * Primary duty cycle, in number of "beats" / "pulse cycle".
+   *
+   * Valid range: [0, beats_per_cycle)
+   *
+   * Note: the raw value written to the `A_*` bitfield in each PWM channel's
+   *       `DUTY_CYCLE_*` CSR is in units of "phase counter ticks", not "beats".
+   *       However, the hardware only takes into account the first `DC_RESN` + 1
+   *       MSBs of the raw duty cycle value to determine the number of "beats"
+   *       for a given duty cycle. To make this configuration easier, the
+   *       software manages the conversion from "beats_per_cycle" to
+   *       "phase_counter_ticks". under the hood.
+   */
+  uint16_t duty_cycle_a;
+  /**
+   * Secondary duty cycle, in number of "beats" / "pulse cycle", that is only
+   * relevant in heartbeat and blink modes.
+   *
+   * Valid range: [0, beats_per_cycle)
+   *
+   * Note: above note for `duty_cycle_a` applies here too.
+   */
+  uint16_t duty_cycle_b;
+  /**
+   * The operation mode to configure the channel in, see `dif_pwm_mode_t`.
+   */
+  dif_pwm_mode_t mode;
+  /**
+   * The polarity to configure the channel in, see `dif_pwm_polarity_t`.
+   */
+  dif_pwm_polarity_t polarity;
+  /**
+   * One of two blink parameters that only impact the "Heartbeat" and "Blink"
+   * operation modes.
+   *
+   * The meaning of this parameter is different based on the operation mode:
+   *     - Heartbeat mode: determines the number of "pulse cycles" between
+   *                       increments/decrements to the duty cycle.
+   *
+   *     - Blink mode: determines the number of "pulse cycles" to pulse at duty
+   *                   cycle A, before switching to duty cycle B.
+   */
+  uint16_t blink_parameter_x;
```

Comment:
TBH I would have made this into a `u16[2]` to hel make the comment shorter but it's w/e